### PR TITLE
subsume wrap under create

### PR DESCRIPTION
**Pull request by madhur-tandon**

### CODE REVIEW

File: src/pyscript/plugins/create.py [hunk 1]
```diff
@@ -20,25 +27,111 @@ def create(
...
+
+    if (title or output or show or command) and (not wrap):
         raise cli.Abort(
-            f"A directory called {app_name} already exists in this location."
+            """`--title`, `--output/-o`, `--show` and `--command/-c`
+            are meant to be used with `--wrap/-w`"""
         )
 
+    if not wrap:
+        try:
```
Comment (FabioRosado) [inline src/pyscript/plugins/create.py:78]: Should the try be on line 84?

File: src/pyscript/plugins/create.py [hunk 2]
```diff
@@ -20,25 +27,111 @@ def create(
...
+            if not app_description:
+                app_description = typer.prompt("App description")
+            if not author_name:
+                author_name = typer.prompt("Author name")
+            if not author_email:
+                author_email = typer.prompt("Author email")
+            create_project(
+                str(app_or_file_name),
```
Comment (FabioRosado) [inline src/pyscript/plugins/create.py:86]: If I am reading the code correct `app_or_file_name` may still be None here right? If that's the case we probably don't want to str a None value 🤔

File: src/pyscript/plugins/create.py [hunk 3]
```diff
@@ -20,25 +27,111 @@ def create(
...
+                template_name="wrap.html",
+                pyscript_version=pyscript_version,
+            )
+        if output:
+            if show:
+                console.print("Opening in web browser!")
+                webbrowser.open(f"file://{output.resolve()}")
+            if remove_output:
+                time.sleep(1)
+                output.unlink()
```
Comment (FabioRosado) [inline src/pyscript/plugins/create.py:133]: Just a thought, perhaps this logic could be placed inside a function? This if/else is a bit hard to follow

File: src/pyscript/plugins/create.py [hunk 4]
```diff
@@ -20,25 +27,111 @@ def create(
...
+
```
Comment (mchilvers) [inline src/pyscript/plugins/create.py:134]: Maybe extract this method into a coupla smaller ones?

**Review by mchilvers — APPROVED** [review]

File: src/pyscript/plugins/create.py (hunk 2 above)
Comment (madhur-tandon) [reply src/pyscript/plugins/create.py:86]: It won't matter since there are checks that prevent this + if you don't pass anything i.e. if it is None, then `typer` has safeguards aka it will prompt that you need to enter an argument

File: src/pyscript/plugins/create.py [hunk 5]
```diff
@@ -20,25 +27,111 @@ def create(
...
+    command: Optional[str] = typer.Option(
+        None,
+        "-c",
+        "--command",
+        help="If provided, embed a single command string. Meant to be used with `--wrap`",
+    ),
+    title: Optional[str] = typer.Option(
+        None, help="Add title to HTML file. Meant to be used with `--wrap`"
+    ),
+    show: Optional[bool] = typer.Option(
```
[the remark this answers is not in the document]
Comment (madhur-tandon) [reply src/pyscript/plugins/create.py:52]: That was for using `--no-view` instead of `--silent`. Do we wanna do this for `show` as well?

File: src/pyscript/plugins/create.py (hunk 3 above)
Comment (madhur-tandon) [reply src/pyscript/plugins/create.py:133]: Yeah, thanks!

File: README.md [hunk 6]
```diff
@@ -109,3 +71,41 @@ The following files will be created:
...
+
+```shell
+$ pyscript create --wrap <filename.py> -o <another_filename.html>
+```
+
+- ##### Open the script inside the browser using the `--show` option
+
+```shell
+$ pyscript create --wrap <filename.py> --show
+```
```
Comment (fpliger) [inline README.md:94]: -1 on this. This should be an option under the `run` cmd and if it means we are just showing the file, without running a static server it'll be prone to a bunch of possible errors that would probably frustrate users

File: README.md [hunk 7]
```diff
@@ -109,3 +71,41 @@ The following files will be created:
...
+
+- ##### Set a title for the browser tab
+
+You can set the title of the browser tab with the `--title` option:
+
+```shell
+$ pyscript create --wrap <filename.py> --title "My cool app!"
+```
```
Comment (fpliger) [inline README.md:102]: In theory we were to use the `name` field in the toml to set the page title. Is thi doing something different? I think it'd be good to have a bit of feature discovery discussion about it to try and find the best option

File: src/pyscript/plugins/create.py [hunk 8]
```diff
@@ -1,15 +1,61 @@
...
 import typer
 
-from pyscript import LATEST_PYSCRIPT_VERSION, app, cli, plugins
-from pyscript._generator import create_project
+from pyscript import LATEST_PYSCRIPT_VERSION, app, cli, console, plugins
+from pyscript._generator import create_project, file_to_html, string_to_html
+
+
+def wrap_file(title, output, command, show, app_or_file_name, pyscript_version):
+    title = title or "PyScript App"
```
Comment (fpliger) [inline src/pyscript/plugins/create.py:13]: Docstring

File: README.md (hunk 6 above)
Comment (madhur-tandon) [reply README.md:94]: This is also borrowed from the previous functionality of `pyscript wrap`. It only shows the file on the browser yes, without the server.

Now that you mention this, I think we can remove this `--show` option altogether and let `pyscript run` handle this, as it does already.

File: src/pyscript/plugins/create.py [hunk 9]
```diff
@@ -1,15 +1,61 @@
...
+
+    # Derive the output path if it is not provided
+    remove_output = False
+    if output is None:
+        if command and show:
+            output = Path("pyscript_tmp.html")
+            remove_output = True
```
Comment (fpliger) [inline src/pyscript/plugins/create.py:20]: I don't think I understand this. Can you explain what happens here and why we need to remove the output?

File: src/pyscript/plugins/create.py [hunk 10]
```diff
@@ -1,15 +1,61 @@
...
+            raise cli.Abort("Must provide an output file or use `--show` option")
+    if app_or_file_name is not None:
+        file_to_html(
+            app_or_file_name,
+            title,
+            output,
+            template_name="wrap.html",
+            pyscript_version=pyscript_version,
+        )
+    if command:
```
Comment (fpliger) [inline src/pyscript/plugins/create.py:34]: This is probably me being picky but wouldn't be better to do an elif and raise an error in the else case? (because you either have a command or a file)

File: src/pyscript/plugins/create.py [hunk 11]
```diff
@@ -1,15 +1,61 @@
...
+            command,
+            title,
+            output,
+            template_name="wrap.html",
+            pyscript_version=pyscript_version,
+        )
+    if output:
+        if show:
+            console.print("Opening in web browser!")
+            webbrowser.open(f"file://{output.resolve()}")
```
Comment (fpliger) [inline src/pyscript/plugins/create.py:45]: -1 on this for the reasons above

File: src/pyscript/plugins/create.py (hunk 9 above)
Comment (madhur-tandon) [reply src/pyscript/plugins/create.py:20]: This is if we are using `pyscript create --wrap -c 'print("Hi!")' --show` let's say aka a command line string with the
`--show` option, then we create a temporary HTML file, open it in the browser (without the server), and then delete it later.

This is also borrowed / carried from what `pyscript wrap` was doing earlier.

File: src/pyscript/plugins/create.py [hunk 12]
```diff
@@ -20,25 +66,76 @@ def create(
...
         )
 
+    if not wrap:
+        if not app_description:
+            app_description = typer.prompt("App description")
+        if not author_name:
+            author_name = typer.prompt("Author name")
+        if not author_email:
+            author_email = typer.prompt("Author email")
+        try:
```
Comment (fpliger) [inline src/pyscript/plugins/create.py:123]: I do think we need to revisit these options above in a different PR/Discussion but, assuming we keep things as is, why we don't use them for wrap?

File: src/pyscript/plugins/create.py (hunk 11 above)
Comment (madhur-tandon) [reply src/pyscript/plugins/create.py:45]: Yeah, I think we should remove `--show` from here altogether.

Once again, this is the reminiscent of the past of the `pyscript wrap` functionality.

File: src/pyscript/plugins/create.py (hunk 2 above)
Comment (FabioRosado) [reply src/pyscript/plugins/create.py:125]: Do you mean that `create_project` checs if the `app_or_file_name` is the string `None` and remove it? Because that's what will happen correct?

**Review by fpliger — CHANGES_REQUESTED** [review]
Left comments. In general I'd expect a different behaviour than what's implemented in the PR.

As an user, I'd like for the cmd to create a project in the same structure `create` by itself does, ensuring best practices and creating a project that the users in the right track

File: src/pyscript/plugins/create.py (hunk 12 above)
Comment (madhur-tandon) [reply src/pyscript/plugins/create.py:123]: because the idea of the original `pyscript wrap` was just to spit out an HTML file. And not really a project folder.

I haven't changed the functionality, merely borrowed it and subsumed it under `create`

File: src/pyscript/plugins/create.py (hunk 2 above)
Comment (madhur-tandon) [reply src/pyscript/plugins/create.py:125]: No, not the `create_project` function, but the `pyscript create` command, that uses typer.

Comment (FabioRosado) [reply src/pyscript/plugins/create.py:125]: I might be missing something, we set `app_or_file_name` with:

```
app_or_file_name: Optional[Path] = typer.Argument(
        None, help="The name of your new app or path to an input .py script"
    ),
```

So that means this could be none since that's the default? Because we then have:

```
if not app_or_file_name and not command:
```

Which makes me assume that app_or_file_name can indeed be none? 🤔

Comment (madhur-tandon) [reply src/pyscript/plugins/create.py]: Ideally, typer was handling for us because we used `typer.Argument` instead of `typer.Option`

But, I did come across a case when we use `-w` and `-c` together i.e. `wrap` and `command`.

In this case, indeed `app_or_file_name` is kinda `None`, I have handled it now.

File: src/pyscript/plugins/create.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (madhur-tandon) [reply src/pyscript/plugins/create.py]: We have removed `show` now.

File: README.md [hunk 13]
```diff
@@ -109,3 +71,41 @@ The following files will be created:
...
+#### Use --wrap to embed a python file OR a command string
+
+- ##### Embed a Python script into a PyScript HTML file
+
+```shell
+$ pyscript create --wrap <filename.py>
+```
+
+This will generate a file called `<filename.html>` by default.
+This can be overwritten with the `-o` or `--output` option:
```
[the remark this answers is not in the document]
Comment (madhur-tandon) [reply README.md]: Now, it is a project folder with the generated `filename.html` inside it :)
So it no longer conforms to what was present in the earlier implementation.

We now create a project folder always!

File: README.md (hunk 6 above)
Comment (madhur-tandon) [reply README.md]: Removed `show` altogether as discussed.

File: README.md (hunk 7 above)
Comment (madhur-tandon) [reply README.md]: Removed `title` as well along with `show` as of now.

For the `wrap` option, either a python file or a command string, the default title is `PyScript App` as of now.

File: src/pyscript/plugins/create.py (hunk 8 above)
Comment (madhur-tandon) [reply src/pyscript/plugins/create.py]: Removed this altogether because of the new implementation.

File: src/pyscript/plugins/create.py (hunk 9 above)
Comment (madhur-tandon) [reply src/pyscript/plugins/create.py]: No longer needed since implementation has changed and `show` is also removed.

File: src/pyscript/plugins/create.py (hunk 10 above)
Comment (madhur-tandon) [reply src/pyscript/plugins/create.py]: Simplified a bit with the new implementation.

File: src/pyscript/plugins/create.py (hunk 11 above)
Comment (madhur-tandon) [reply src/pyscript/plugins/create.py]: Removed `show` altogether.

File: src/pyscript/plugins/create.py (hunk 12 above)
Comment (madhur-tandon) [reply src/pyscript/plugins/create.py]: We are now using them for `wrap` as well in the new implementation :D

File: src/pyscript/_generator.py [hunk 14]
```diff
@@ -140,26 +156,52 @@ def create_project(
...
-        )
+    if not wrap:
+        output = app_dir / "index.html"
+        if project_type == "app":
+            template = "basic.html"
+        elif project_type == "plugin":
+            template = "plugin.html"
+        else:
+            raise ValueError(
+                f"Unknown project type: {project_type}. Valid values are: 'app' and 'plugin'"
```
Comment (FabioRosado) [inline src/pyscript/_generator.py:167]: Just a comment for a follow up perhaps but I think it would be good to put this in a consta and get this logic from somewhere that way we just need to add a new type and both the error message and template logic will stay the same

**Review by FabioRosado — APPROVED** [review]

**Review by tedpatrick — APPROVED** [review]

File: src/pyscript/_generator.py [hunk 15]
```diff
@@ -111,12 +111,15 @@ def file_to_html(
```
Comment (mchilvers) [inline src/pyscript/_generator.py:112]: Maybe the method below could be factored into 2 internal methods? One for the regular create and one for the wrapping case? Then there would be a single switch for 'if wrap'...

File: README.md [hunk 16]
```diff
@@ -109,3 +71,45 @@ The following files will be created:
...
+```shell
+$ pyscript create --wrap <filename.py>
+```
+
+This will generate a project i.e. a new directory named `filename` under the current directory.
+
+Similar to the above, interactive prompts will further ask for metadata information.
+
+The following files will be created:
+- `filename.html`: embeds the code of `filename.py` in a `<py-script>` tag.
```
Comment (fpliger) [inline README.md:88]: -0.5 on this one. This makes the project layout a bit less predictable which is less optimal to the user.

I'd prefer to have the project creation output to be standard and predicable: index.html, main.py and pyscript.toml ; no matter what's the name of the input

File: README.md [hunk 17]
```diff
@@ -109,3 +71,45 @@ The following files will be created:
...
+- `pyscript.toml`: project metadata and config file
+
+This can be overridden with the `-o` or `--output` option:
+
+```shell
+$ pyscript create --wrap <filename.py> -o <another_filename.html>
```
Comment (fpliger) [inline README.md:94]: See my comment above. I wouldn't  do this since I think it just adds complications. If it's something users ask in the future we can get back to it...

Also, we wouldn't create a `py` file? Would the code be inline?

File: README.md [hunk 18]
```diff
@@ -109,3 +71,45 @@ The following files will be created:
...
+
+```shell
+$ pyscript create --wrap -c 'print("Hello World!")' -o <output_filename.html>
+```
+
+This will generate a project i.e. a new directory named `output_filename` under the current directory.
+
+Similar to the above, interactive prompts will further ask for metadata information.
+
+The following files will be created:
```
Comment (fpliger) [inline README.md:113]: same comments as above....  it may be ok to create an inline option too where we don't create the python file and put the code within a `<script type="pyodide">` tag but that should be an option.

In general, we want to always try to encourage best practices

**Review by fpliger — COMMENTED** [review]
Before we merge, it'd be good to align on what's the right behavior. Left some comments and I think we should align on these before we merge...

File: README.md (hunk 16 above)
Comment (madhur-tandon) [reply README.md:88]: I see, so we don't want the python code to be inline between `<py-script>` tags and want it to be via
`<py-script src=...>` instead?

File: README.md (hunk 17 above)
Comment (madhur-tandon) [reply README.md:94]: Yes, `wrap` assumes that the code is inline. Similar to the comment above.

It is basically borrowing the behaviour of `wrap` from the past aka `inline` code and blending it with the behaviour of creating a project folder for it.

File: README.md (hunk 18 above)
Comment (madhur-tandon) [reply README.md:113]: Yeah, older `wrap` always made it inline.
And `create` used to create project folders.

So the current behaviour is the blend of both.

Asking again as asked in the first comment, so we don't want inline code ideally?
aka we want it to be via `<py-script src=...>` instead? Aka we do need the python file for `wrap`?

I am asking for a confirmation here since `wrap` was used to embed it as inline.

File: README.md (hunk 16 above)
Comment (madhur-tandon) [reply README.md]: Done!

File: README.md (hunk 17 above)
Comment (madhur-tandon) [reply README.md:96]: Done!

File: README.md (hunk 18 above)
Comment (madhur-tandon) [reply README.md:118]: Done!

File: README.md [hunk 19]
```diff
@@ -109,3 +71,52 @@ The following files will be created:
...
+- `pyscript.toml`: project metadata and config file
+- `main.py`: contains code of `filename.py`
+
+This can be overridden with the `-o` or `--output` option:
+
+```shell
+$ pyscript create --wrap <filename.py> -o <another_filename.html>
+```
+
+i.e. the HTML file created in the above directory will now be named `another_filename.html`
```
Comment (fpliger) [inline README.md:99]: This is not a blocker for the PR (we can fix in a follow up PR) but I'm trying to understand the behavior here. Does this mean that if the users specifices `-o`, we'll create a folder and will have everything in a new single file?

Looking at the test for this seems like, in the case above, we'd have:

* a new `another_filename` folder
* a new file `another_filename\another_filename.html`
* a new `another_filename\main.py` file
* a new `another_filename\pyscript.toml` (test is not checking that but I'm assuming)

Is that correct?

**Review by fpliger — APPROVED** [review]
Overall LGTM. Think there might be a couple of details to verify (and maybe fix later) but 🚀 !!

File: README.md (hunk 19 above)
Comment (madhur-tandon) [reply README.md:99]: Yeah, that's correct except that name of folder is `filename` and not `another_filename`

Comment (fpliger) [reply README.md:99]: ah, I see. Ok. I think it's ok for now... we can sit down and try to make a thought through decision later. Mainly I'm afraid there might be some confusing scenarios..

Comment (madhur-tandon) [reply README.md:99]: Ok, merging for now...
